File: code/ch09/9.2.1.2.linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL || head->next == NULL)
            return false;
        
        ListNode *p1List = head;
        ListNode *p2List = head;
        
        while(p2List != NULL && p2List->next != NULL){
            p1List = p1List->next;
            p2List = p2List->next->next;
            
            if (p1List == p2List)
                return true;
        }
        
        return false;
    }
};

File: code/ch09/9.2.1.2.linked_list_cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null)
            return false;

        ListNode fast = head, slow = head;
        while (fast.next != null) {
            fast = fast.next;
            if (fast == slow)
                return true;
            else {
                if (fast.next == null)
                    return false;
                fast = fast.next;
                slow = slow.next;
            }
        }
        return false;
    }
}